src/include/migraphx/permutation.hpp

```diff
@@ -66,6 +66,10 @@ MIGRAPHX_EXPORT std::vector<int64_t> invert_permutation(const std::vector<int64_
 MIGRAPHX_EXPORT std::vector<int64_t> find_permutation(const shape& s);
 MIGRAPHX_EXPORT std::vector<int64_t> find_permutation(const std::vector<shape>& shapes);
 
+/// Normalize the shapes so the order of dimensions will be in the order it is
+/// in memory as much as possible.
```

Comment:
how does it help with performance? it would still be permutated and strided in memory.

Reply:
>  it would still be permutated and strided in memory.

No it wont since the dimensions are now in the same order as they are in memory. Say we have a transposed shape like this: `{4, 3, 8, 8}, {192, 1, 24, 3}` which has a permutation of `{0, 2, 3, 1}`(ie NHWC). After `normalize_permutation` it will be `{4, 8, 8, 3}, {192, 24, 3, 1}` which is standard shape. So the order of elements in the shape is the same as the order of elements in memory so there wont be any strided reads or writes.

Reply:
I think this would also change input's memory layout if it is broadcasted.

Reply:
> If let's say, it is a pointwise binary operator and one of the input is transposed but the other one is not.

For mixed layouts, it may not still not be optimal(where usually we would need to copy to LDS to do the transpose), but it will still be better then before.

It takes in the output in account as well. So if the output is also transposed so only one tensor will be used as transposed instead of two transposed tensors. Or if one input is NHWC and another CHWN, we can have at least one input be non-transposed rather then neither.

Reply:
> I think this would also change input's memory layout if it is broadcasted.

It never changes the way its layout'd in memory. Also `find_permutation` skips using broadcasted shapes since those can't be inferred accurately due to the 0 strides.

Reply:
>  Also find_permutation skips using broadcasted shapes since those can't be inferred accurately due to the 0 strides. 

find_permutation would skip it but `normalize_permutation` would still transform it.